### PR TITLE
#3 readme added

### PR DESCRIPTION
**Pull request by mrChex**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,263 @@
...
+Result = graphql:execute(graphql_example:schema(), Document, InitValue, Context),
+
+io:format("Result: ~p~n", [Result]).  % #{data => [{<<"name">>, <<"world">>}]}
+```
+
+## Types include
+
+We provide macros for simple types define. Include: `-include_lib("graphql/include/types.hrl").`
+
+All value for types may be `null` except `?NON_NULL` and ObjectType
```
Comment (metallicolis) [inline README.md]: can be null

File: README.md [hunk 2]
```diff
@@ -1 +1,263 @@
...
+where:
+
+- ObjectValue - result of parent resolver (or InitialValue for query root)
+- ArgumentsValues - map of arguments. When argument not provided in query - value is 'null'
+- Context - context from parent resolver.
+
+Resolver can return resolved value witch passed to scalar serialize function if type is scalar or as ObjectValue
+for each field in child object type
+ 
+Resolver can alse overwrite context for his childrens. In this case resolver may look like:
```
Comment (metallicolis) [inline README.md]: else?

File: README.md [hunk 3]
```diff
@@ -1 +1,263 @@
...
+  - [X] Fragments and inline fragments in queries
+  - [X] Custom types
+  - [ ] Parallel execution
+  - [ ] Return maps option
+- [X] Mutations
+- [ ] Subscriptions
+- [X] Introspection
+- [ ] Directives
+- [ ] "Compile" and validate schema before app start
+- [ ] GraphQL Guard (may be another app)
```
Comment (metallicolis) [inline README.md]: maybe

File: README.md [hunk 4]
```diff
@@ -1 +1,263 @@
-# Erlang GraphQL [![Build Status](https://travis-ci.org/graphql-erlang/graphql.svg?branch=master)](https://travis-ci.org/graphql-erlang/graphql)
\ No newline at end of file
+# Erlang GraphQL [![Build Status](https://travis-ci.org/graphql-erlang/graphql.svg?branch=master)](https://travis-ci.org/graphql-erlang/graphql)
+
+An Erlang implementation of Facebook's GraphQL.
+
+This is the core GraphQL query parsing and execution engine whose goal is to be transport, server and datastore agnostic.
```
Comment (jrtidev) [inline README.md]: whose ==>which

File: README.md [hunk 5]
```diff
@@ -1 +1,263 @@
...
+
+
+## Status
+
+This is a work in progress, here's todo list:
```
Comment (jrtidev) [inline README.md:10]: This is my to-do list and table of content:

File: README.md [hunk 6]
```diff
@@ -1 +1,263 @@
...
+  - [X] Variables
+  - [X] Fragments and inline fragments in queries
+  - [X] Custom types
+  - [ ] Parallel execution
+  - [ ] Return maps option
+- [X] Mutations
+- [ ] Subscriptions
+- [X] Introspection
+- [ ] Directives
+- [ ] "Compile" and validate schema before app start
```
Comment (jrtidev) [inline README.md]: "Compile" and validate schema before start the app

File: README.md [hunk 7]
```diff
@@ -1 +1,263 @@
...
+}).
+
+query() -> graphql:objectType(<<"QueryRoot">>, <<"Example Query">>, #{
+  <<"greatings">> => #{
+    type => ?STRING,
+    description => <<"Greating">>,
+    args => #{
+      <<"name">> => #{
+        type => ?NON_NULL(?STRING),
+	description => <<"The name of who you'd like to great.">>
```
Comment (jrtidev) [inline README.md]: "The name of who you'd like to great"

File: README.md [hunk 8]
```diff
@@ -1 +1,263 @@
...
+      maps:get(<<"name">>, Args)
+    end
+  }
+}).
+```
+
+In handler - call query for schema with optional options:
+```erlang
+Document = <<"query($name: String!) { greatings(name: $name) }">>,  % or list
+Context = #{},  % you can pass request through context or whatever you want use in resolver acress the schema
```
Comment (jrtidev) [inline README.md:80]: you can pass request through the context or whatever you want use in resolver to access the schema

File: README.md [hunk 9]
```diff
@@ -1 +1,263 @@
...
+},
+
+Result = graphql:execute(graphql_example:schema(), Document, InitValue, Context),
+
+io:format("Result: ~p~n", [Result]).  % #{data => [{<<"name">>, <<"world">>}]}
+```
+
+## Types include
+
+We provide macros for simple types define. Include: `-include_lib("graphql/include/types.hrl").`
```
Comment (jrtidev) [inline README.md]: define =>definition

File: README.md (hunk 1 above)
Comment (jrtidev) [inline README.md]: value => values
may => can

File: README.md [hunk 10]
```diff
@@ -1 +1,263 @@
...
+
+`?FLOAT` and `?INT` - represent float and integer. 
+
+`?LIST(OfType)` - represent list where `OfType` - inner list type. Example: `?LIST(?INT)` - `[Int]`
+
+`?NON_NULL(OfType)` - non nullabel type - wrapper over nullable type. Example: `?NON_NULL(?LIST(?INT))` - `[Int]!`
+
+`?STRING` - binary string
+
+`?UNION(Name, Description, PossibleTypes, ResolveTypeFun)` - type can be one of listed in PossibleTypes. PossibleTypes must be list of ObjectTypes.
```
Comment (jrtidev) [inline README.md]: type can be one of the listed in

File: README.md [hunk 11]
```diff
@@ -1 +1,263 @@
...
+
+union_custom() -> ?UNION(<<"UnionCastomResolve">>, <<"Used custom type resolver">>, [
+  fun one/0,
+  fun two/0
+], fun(Value, _ThisTypeDefinition)->
+  Type = case maps:get(<<"one">>, Value, undefined) of
+    undefined -> fun two/0;
+    _ -> fun one/0
+  end,
+  {Type, Value}  % resolver must return tuple {ResolvedType, UnwrappedValue}
```
Comment (jrtidev) [inline README.md]: resolver should return tuple

File: README.md [hunk 12]
```diff
@@ -1 +1,263 @@
...
+end).
+```
+
+Object Type definition:
+```erlang
+test_unions() -> graphql:objectType(<<"TestUnions">>, <<>>, #{
+  <<"default_resolver">> => #{
+    type => ?LIST(fun unions_default/0),
+    
+    % default resolver just compare first elem in tuple with possibleTypes in union
```
Comment (jrtidev) [inline README.md]: compare =>compaers

File: README.md [hunk 13]
```diff
@@ -1 +1,263 @@
...
+
+The most basic components of a GraphQL schema are object types, which just represent a kind
+of object you can fetch from your service, and what fields it has.
+
+Creating object type example:
+```erlang
+
+graphql:objectType(<<"Name">>, <<"Description">>, #{
+  <<"field_name">> => #{
+    type => ?STRING,  % what type of resolved value
```
Comment (jrtidev) [inline README.md]: what type of value returned by resolver

File: README.md [hunk 14]
```diff
@@ -1 +1,263 @@
...
+    description => <<"Human redable description">>,  % optional
+    isDeprecated => true, % optional. Default: false
+    deprecationReason => <<"Human redable deprecation reason">>, % used when isDeprecated => true for introspection tools
+    args => #{  % map of field arguments
```
Comment (jrtidev) [inline README.md]: mapping of the field arguments

File: README.md [hunk 15]
```diff
@@ -1 +1,263 @@
...
+        type => ?INT
+        % optional: description, isDeprecated        
+      }
+    },
+    resolver => fun() -> <<"resolved string">>
+  }
+})
+```
+
+Resolver - function for resolving field value. Arity and arguments passing:
```
Comment (jrtidev) [inline README.md]: function that resolves field values. Arity and arguments passing:

File: README.md [hunk 16]
```diff
@@ -1 +1,263 @@
...
+  {arity, 0} -> Resolver();
+  {arity, 1} -> Resolver(ObjectValue);
+  {arity, 2} -> Resolver(ObjectValue, ArgumentValues);
+  {arity, 3} -> Resolver(ObjectValue, ArgumentValues, Context)
+end.
+```
+
+where:
+
+- ObjectValue - result of parent resolver (or InitialValue for query root)
```
Comment (jrtidev) [inline README.md]: result of the parent resolver

File: README.md [hunk 17]
```diff
@@ -1 +1,263 @@
...
+- ArgumentsValues - map of arguments. When argument not provided in query - value is 'null'
+- Context - context from parent resolver.
+
+Resolver can return resolved value witch passed to scalar serialize function if type is scalar or as ObjectValue
```
Comment (jrtidev) [inline README.md]: Resolver can return resolved value which was passed to scalar serializing function if type is scalar or as ObjectValue

File: README.md (hunk 2 above)
Comment (jrtidev) [inline README.md]: Resolver can also overwrite context of his children. In this case resolver can looks like:

File: README.md [hunk 18]
```diff
@@ -1 +1,263 @@
...
+  name => 'IntCustom',  % atom
+  ofType => null,  % used for nested data types
+  description => <<
+    "The `Int` scalar type represents non-fractional signed whole numeric ",
+    "values. Int can represent values between -(2^53 - 1) and 2^53 - 1 since ",
+    "represented in JSON as double-precision floating point numbers specified ",
+    "by [IEEE 754](http://en.wikipedia.org/wiki/IEEE_floating_point)."
+  >>,
+
+  % this function call for serialize resolved value
```
Comment (jrtidev) [inline README.md]: you can call this function to serialize resolved value

File: README.md [hunk 19]
```diff
@@ -1 +1,263 @@
...
+  % Serialize(Result, FieldType, Context);
+  serialize => fun serialize/3,
+  
+  % this function call for parsing variable value
```
Comment (jrtidev) [inline README.md]: you can call this function to parse value of the variable

File: README.md [hunk 20]
```diff
@@ -1 +1,263 @@
...
+  % ParseValue(maps:get(VariableName, VariableValues, null), ArgumentType);
+  parse_value => fun parse_value/2,
+  
+  % this function call for parsing value defined in query
```
Comment (jrtidev) [inline README.md]: you can call this function to parse the value defined in query

**Review by jrtidev — CHANGES_REQUESTED** [review]

File: README.md (hunk 5 above)
Comment (mrChex) [reply README.md:10]: Но тут речь о том, что библиотека все еще в работе и вот что еще нужно закончить

File: README.md (hunk 8 above)
Comment (mrChex) [reply README.md:80]: across - контекст - это то что передается во все резолверы по всей схеме

File: README.md (hunk 14 above)
Comment (mrChex) [reply README.md]: map - data structure

**Review by metallicolis — APPROVED** [review]

**Pull request merged**